auto_nag/multi_autofixers.py

```diff
@@ -0,0 +1,145 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import argparse
+from collections import ChainMap, defaultdict
+from typing import Any, Callable, Counter, Dict, List, Type
+
+from auto_nag import logger, utils
+from auto_nag.bzcleaner import BzCleaner
+from auto_nag.nag_me import Nag
+
+ToolsChanges = Dict[Type[BzCleaner], Any]
+
+
+class UnexpectedToolsError(Exception):
+    """Unexpected tools appear in merge function"""
+
+    def __init__(self, tools: List[Type[BzCleaner]]) -> None:
+        """Constructor
+
+        Args:
+            tools: the tools that change the same field.
+        """
+        super().__init__()
+        self.tools = tools
+
+    def __str__(self):
+        return (
+            "Error: merge function does not support merge fields "
+            f"from '{utils.english_list(self.tools)}'"
+        )
+
+
+class MissingMergeFunctionError(Exception):
+    """Merge function is missing"""
+
+    def __init__(self, field: str) -> None:
+        """Constructor
+
+        Args:
+            field: the field that is missing a merge function.
+        """
+        super().__init__()
+        self.field = field
+
+    def __str__(self):
+        return f"Error: missing merge function for '{self.field}'"
+
+
+class MultiAutoFixers:
+    "Merge changes from multiple tools and apply them to Bugzilla at once"
+
+    def __init__(
+        self, *tools: BzCleaner, **merge_functions: Callable[[ToolsChanges], dict]
+    ):
+        """Constructor
+
+        Args:
+            *tools: tools to merged their changes.
+            **merge_functions: functions to merged the fields when multiple
+                tools are changing the same fields. The key should be the field
+                name.
+        """
+        super().__init__()
+        for tool in tools:
+            if isinstance(tool, Nag):
+                logger.warning(
+                    "%s implements Nag, however, nag emails will not be merged",
+                    type(tool),
+                )
+        self.tools = tools
+        self.merge_functions = merge_functions
+        self.is_dryrun: bool = True
+
+    def description(self):
+        """Return a description of the tool"""
+        return "Grouped changes for the following tools: " + self.name()
+
+    def name(self):
+        """Return the name of the tool"""
+        return utils.english_list([tool.name() for tool in self.tools])
+
+    def get_args_parser(self):
+        """Get the argumends from the command line"""
```

Comment:
Typo

Reply:
Nice catch. This typo has two clones that have been living in the repo for 4 years :)

Fixed in b84ea58cdf07dce2eb0abc5604d8f114cb6226d8